pkg/collector/corechecks/servicediscovery/module/impl_linux.go

```diff
@@ -613,6 +614,65 @@ func (s *discovery) updateServicesCPUStats(services []model.Service) error {
 	return nil
 }
 
+func (s *discovery) enrichContainerData(service *model.Service, containers []*agentPayload.Container, pidToCid map[int]string) {
+	id, ok := pidToCid[service.PID]
+	if !ok {
+		return
+	}
+
+	// The tags we look for service name generation, in their priority order.
+	// The map entries will be filled as we go through the containers tags.
+	tagsPriority := []struct {
+		tagName  string
+		tagValue *string
+	}{
+		{"service", nil},
+		{"app", nil},
+		{"short_image", nil},
+		{"kube_container_name", nil},
+		{"kube_deployment", nil},
+		{"kube_service", nil},
+	}
+
+	service.ContainerID = id
+	log.Debugf("Found container id for %v: %v", service.Name, id)
+	for _, c := range containers {
+		if c.Id != id {
+			continue
+		}
+
+		for _, tag := range c.Tags {
+			// Get index of separator between name and value
+			sepIndex := strings.IndexRune(tag, ':')
+			if sepIndex < 0 || sepIndex >= len(tag)-1 {
+				// Malformed tag; we skip it
+				continue
+			}
+
+			for i := range tagsPriority {
+				if tag[:sepIndex] != tagsPriority[i].tagName {
+					// Not a tag we care about; we skip it
+					continue
+				}
+
+				value := tag[sepIndex+1:]
+				tagsPriority[i].tagValue = &value
+				break
+			}
+		}
+
+		for _, tag := range tagsPriority {
+			if tag.tagValue != nil {
+				service.GeneratedName = *tag.tagValue
```

Comment:
> Can we save the generated name in the info cache so that we don't have to do the above computation for every invocation of the endpoint?

I agree with the idea, but in practice, we can't rely on the `generatedName` field being set, because it always will be first by `getService`.